Adapt Sensei Core to the Sensei Co-Teachers functionality.

aaronfc opened the pull request:
### Changes proposed in this Pull Request
* Replace check for author with check for `edit_post` capability of the lesson for Quizzes.

### Testing instructions
* Create a lesson.
* Confirm you can get a response from the `sensei-internal/v1/lesson-quiz/3033?context=edit&_locale=user` API (you can open the editor for the lesson and it will do the request for you).
* With **another user** confirm you get a permission error when accessing the `sensei-internal/v1/lesson-quiz/3033?context=edit&_locale=user` API.

### New hooks
- `sensei_filter_module_terms_by_owner`: Filter that allows to alter the result of the module terms when being filtered checking for the owner.
- `sensei_teacher_meta_box_coteachers_feature`: Action that is expected to render the coteachers meta box content.

<!--
Helpful tips for screenshots:
https://en.support.wordpress.com/make-a-screenshot/
-->
### Screenshot / Video
TBD

aaronfc committed:
Change permission check for Quiz REST API access.

- Instead of checking for the user being the post_author we are checking
  for the user having the edit_post capability for the associated
  lesson.

aaronfc committed:
Add `sensei_teacher_metabox_coteachers_feature` hook.

aaronfc commented on includes/class-sensei-teacher.php:
```diff
@@ -278,6 +278,28 @@ public function teacher_meta_box_content( $post ) {
 
+		// Get `sensei_teacher_metabox_coteachers_feature` contents.
+		ob_start();
+		do_action( 'sensei_teacher_metabox_coteachers_feature', $post );
```
This is the action that Sensei Pro is hooking into.

aaronfc committed:
Refactor: Renaming metabox to meta_box.

- Affecting some comments and also the
  `sensei_teacher_meta_box_coteachers_feature` hook.

aaronfc committed:
Add `sensei_filter_module_terms_by_owner` filter.

- This filter allows module terms to be altered when being filtered by
  owner.

aaronfc committed:
Merge remote-tracking branch 'origin/trunk' into feature/sensei-coteachers

alexsanford commented on includes/class-sensei-modules.php:
```diff
@@ -2492,8 +2492,7 @@ public function filter_terms_by_owner( $terms, $user_id ) {
 
-		return $users_terms;
-
+		return apply_filters( 'sensei_filter_module_terms_by_owner', $users_terms, $terms, $user_id );
```
Could we add a doc comment for this new filter?

aaronfc replied:
Done [34b8f15](https://github.com/Automattic/sensei/pull/5345/commits/34b8f15d5e8ce31d860b1e3ddd1bd9849591909b)

alexsanford commented on includes/class-sensei-teacher.php:
```diff
@@ -278,6 +278,28 @@ public function teacher_meta_box_content( $post ) {
 
+		// Get `sensei_teacher_meta_box_coteachers_feature` contents.
+		ob_start();
+		do_action( 'sensei_teacher_meta_box_coteachers_feature', $post );
```
I don't really like having an action that mentions a specific feature like this. I would prefer an action that is more generic (perhaps `sensei_teacher_meta_box_after`?) and an explicit hook for hiding the upsell (perhaps `sensei_course_coteachers_hide` in keeping with our other [other upsell filters](https://github.com/Automattic/sensei/blob/5f35d440c6ecbf7f5250f8868bd616e85b40389f/includes/class-sensei-lesson.php#L292)?).

aaronfc replied:
Done: [571a998](https://github.com/Automattic/sensei/pull/5345/commits/571a998d7558a854b5ae7a0fed983300e42ffe2f).

aaronfc committed:
Merge remote-tracking branch 'origin/trunk' into feature/sensei-coteachers

aaronfc committed:
Update Co-Teachers upgrade link from pricing to landing.

aaronfc committed:
Split single hook action into action and filter for Co-Teachers.

- Instead of having a single specific hook for the Co-Teachers feature
  we have now two more generic ones.
- A new action hook to add content to the end of the Teachers meta box.
- A new filter hook that enables/disables the Co-Teachers upgrade
  promotional link.

aaronfc committed:
Add doc for `sensei_filter_module_terms_by_owner` filter.

aaronfc committed:
Escape upgrade promo link.

- Escaping with wp_kses.
- Adding target=_blank.

aaronfc committed:
Fix code-style.

aaronfc committed:
Update with `trunk`.

aaronfc commented:
New branch: `add/coteachers-metabox-hook`
New PR: https://github.com/Automattic/sensei/pull/6115
